Add support for bent normal maps for specular occlusion and indirect lighting

LunaCapra opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/9401

Implements bent normal map support for Forward+ and Mobile and derives specular occlusion from it and ambient occlusion when enabled on a material.
Bent normals are also used for indirect lighting when available.

Examples (**SSAO is enabled**):

| `master` | This PR |
|---|---|
| ![screenshot_2](https://github.com/user-attachments/assets/801974ad-be08-45b6-8eda-0a9e36d3fc11)| ![screenshot_0](https://github.com/user-attachments/assets/bd4ee232-7983-44ea-9feb-71d60a7ed90e) |
|![screenshot_6](https://github.com/user-attachments/assets/7c62b19e-ca9a-44eb-846a-4dc87da9bcf5)|![screenshot_4](https://github.com/user-attachments/assets/0634c5f1-2716-41aa-852d-ff0e46c86f97)|
|![screenshot_10](https://github.com/user-attachments/assets/167aa168-a921-4a9e-a506-e2ed78abe6e1)|![screenshot_8](https://github.com/user-attachments/assets/7f229807-8846-46ec-b50e-1ffa5868430e)|

-----

The test project used can be downloaded here: https://github.com/LunaCapra/godot-bent-normals-test

AThousandShips commented on drivers/gles3/shaders/scene.glsl:
```diff
@@ -1368,6 +1368,11 @@ void main() {
 #endif
 
+#if defined(BENT_NORMAL_MAP_USED)
+
```
```suggestion
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -993,6 +994,12 @@ void fragment_shader(in SceneData scene_data) {
 #endif
 
+#if defined(BENT_NORMAL_MAP_USED)
+
```
```suggestion
```

AThousandShips commented on scene/resources/material.cpp:
```diff
@@ -1542,6 +1549,13 @@ void fragment() {)";
 	}
 
+	if (features[FEATURE_BENT_NORMAL_MAPPING]) {
+		code += R"(
+	// Bent Normal Map: Enabled
+)";
+		code += "	BENT_NORMAL_MAP = texture(texture_bent_normal, base_uv).rgb;\n";
```
```suggestion
		code += R"(
	// Bent Normal Map: Enabled
	BENT_NORMAL_MAP = texture(texture_bent_normal, base_uv).rgb;";
)";
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -1080,6 +1087,15 @@ void fragment_shader(in SceneData scene_data) {
 #endif
 
+#ifdef BENT_NORMAL_MAP_USED
+
+	bent_normal_map.xy = bent_normal_map.xy * 2.0 - 1.0;
+	bent_normal_map.z = sqrt(max(0.0, 1.0 - dot(bent_normal_map.xy, bent_normal_map.xy)));
+
+	bent_normal_vector = normalize(tangent * bent_normal_map.x + binormal * bent_normal_map.y + normal * bent_normal_map.z);
+
```
```suggestion
	bent_normal_map.xy = bent_normal_map.xy * 2.0 - 1.0;
	bent_normal_map.z = sqrt(max(0.0, 1.0 - dot(bent_normal_map.xy, bent_normal_map.xy)));

	bent_normal_vector = normalize(tangent * bent_normal_map.x + binormal * bent_normal_map.y + normal * bent_normal_map.z);
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2391,8 +2422,53 @@ void fragment_shader(in SceneData scene_data) {
+	float specular_occlusion = clamp(pow(ndotv + ao, exp2(-16.0 * roughness - 1.0)) - 1.0 + ao, 0.0, 1.0);
+	indirect_specular_light *= ao_base * specular_occlusion;
+#else
+
+	// Apply cone to cone intersection with cosine weighted assumption
+	// <https://blog.selfshadow.com/publications/s2016-shading-course/activision/s2016_pbs_activision_occlusion.pdf>
+	float cos_a_v = sqrt(1.0 - ao);
+    float limited_roughness = max(roughness, 0.01); // Avoid artifacts at really low roughness
+    float cos_a_s = exp2((-log(10.0) / log(2.0)) * limited_roughness * limited_roughness);
+    float cos_b = dot(bent_normal_vector, reflect(-view, normal));
+
+	// Cone intersection
+	float r1 = acos(cos_a_v);
+    float r2 = acos(cos_a_s);
+    float rd = acos(cos_b);
+    float area = 0.0;
+
+    if (rd <= max(r1, r2) - min(r1, r2))
+    {
+        // One cone is enclosed in the other
+        area = M_TAU - M_TAU * max(cos_a_v, cos_a_s);
+    }
+    else if (rd >= r1 + r2)
+    {
+        // No intersection
+        area = 0.0;
+    }
+    else
+    {
+        float diff = abs(r1 - r2);
+        float den = r1 + r2 - diff;
+        float x = 1.0 - clamp((rd - diff) / den, 0.0, 1.0);
+        area = smoothstep(0.0, 1.0, x);
+        area *= M_TAU - M_TAU * max(cos_a_v, cos_a_s);
+    }
+
+	float specular_occlusion = area / (M_TAU * (1.0 - cos_a_s));
+	indirect_specular_light *= ao_base * specular_occlusion;
+
+#endif
```
```suggestion
#ifdef BENT_NORMAL_MAP_USED
	// Apply cone to cone intersection with cosine weighted assumption:
	// https://blog.selfshadow.com/publications/s2016-shading-course/activision/s2016_pbs_activision_occlusion.pdf
	float cos_a_v = sqrt(1.0 - ao);
    float limited_roughness = max(roughness, 0.01); // Avoid artifacts at really low roughness.
    float cos_a_s = exp2((-log(10.0) / log(2.0)) * limited_roughness * limited_roughness);
    float cos_b = dot(bent_normal_vector, reflect(-view, normal));

	// Cone intersection.
	float r1 = acos(cos_a_v);
    float r2 = acos(cos_a_s);
    float rd = acos(cos_b);
    float area = 0.0;

    if (rd <= max(r1, r2) - min(r1, r2)) {
        // One cone is enclosed in the other.
        area = M_TAU - M_TAU * max(cos_a_v, cos_a_s);
    } else if (rd >= r1 + r2) {
        // No intersection
        area = 0.0;
    } else {
        float diff = abs(r1 - r2);
        float den = r1 + r2 - diff;
        float x = 1.0 - clamp((rd - diff) / den, 0.0, 1.0);
        area = smoothstep(0.0, 1.0, x);
        area *= M_TAU - M_TAU * max(cos_a_v, cos_a_s);
    }

	float specular_occlusion = area / (M_TAU * (1.0 - cos_a_s));
	indirect_specular_light *= ao_base * specular_occlusion;
#else // BENT_NORMAL_MAP_USED
	// Apply indirect light AO according to Filament approximation:
	// https://google.github.io/filament/Filament.html#lighting/occlusion/specularocclusion
	float ndotv = clamp(dot(normal, view), 0.0, 1.0);
	float specular_occlusion = clamp(pow(ndotv + ao, exp2(-16.0 * roughness - 1.0)) - 1.0 + ao, 0.0, 1.0);
	indirect_specular_light *= ao_base * specular_occlusion;
#endif // BENT_NORMAL_MAP_USED
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl:
```diff
@@ -800,6 +801,13 @@ void main() {
 #endif
 
+#if defined(BENT_NORMAL_MAP_USED)
+
+	vec3 bent_normal_vector;
+	vec3 bent_normal_map = vec3(0.5);
+
```
```suggestion
	vec3 bent_normal_vector;
	vec3 bent_normal_map = vec3(0.5);
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl:
```diff
@@ -893,6 +901,15 @@ void main() {
 #endif
 
+#ifdef BENT_NORMAL_MAP_USED
+
+	bent_normal_map.xy = bent_normal_map.xy * 2.0 - 1.0;
+	bent_normal_map.z = sqrt(max(0.0, 1.0 - dot(bent_normal_map.xy, bent_normal_map.xy)));
+
+	bent_normal_vector = normalize(tangent * bent_normal_map.x + binormal * bent_normal_map.y + normal * bent_normal_map.z);
+
```
```suggestion
	bent_normal_map.xy = bent_normal_map.xy * 2.0 - 1.0;
	bent_normal_map.z = sqrt(max(0.0, 1.0 - dot(bent_normal_map.xy, bent_normal_map.xy)));

	bent_normal_vector = normalize(tangent * bent_normal_map.x + binormal * bent_normal_map.y + normal * bent_normal_map.z);
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl:
```diff
@@ -1792,8 +1821,22 @@ void main() {
+	direct_specular_light *= ao;
+
+#ifdef BENT_NORMAL_MAP_USED
+
+	// Simplified bent normal occlusion
+	float cos_b = max(dot(reflect(-view, normal), bent_normal_vector), 0.0);
+	float specular_occlusion = clamp((ao_base - (1.0 - cos_b)) / roughness, 0.0, 1.0);
+	specular_occlusion = mix(specular_occlusion, cos_b * (1.0 - ao_base), roughness);
+	indirect_specular_light *= specular_occlusion;
+
+#else
+
+	indirect_specular_light *= ao_base;
+
```
```suggestion
	// Simplified bent normal occlusion.
	float cos_b = max(dot(reflect(-view, normal), bent_normal_vector), 0.0);
	float specular_occlusion = clamp((ao_base - (1.0 - cos_b)) / roughness, 0.0, 1.0);
	specular_occlusion = mix(specular_occlusion, cos_b * (1.0 - ao_base), roughness);
	indirect_specular_light *= specular_occlusion;
#else
	indirect_specular_light *= ao_base;
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl:
```diff
@@ -22,13 +23,13 @@
 #endif
 
-#if !defined(TANGENT_USED) && (defined(NORMAL_MAP_USED) || defined(LIGHT_ANISOTROPY_USED))
+#if !defined(TANGENT_USED) && ((defined(NORMAL_MAP_USED) || defined(BENT_NORMAL_MAP_USED)) || defined(LIGHT_ANISOTROPY_USED))
```
```suggestion
#if !defined(TANGENT_USED) && (defined(NORMAL_MAP_USED) || defined(BENT_NORMAL_MAP_USED) || defined(LIGHT_ANISOTROPY_USED))
```

LunaCapra commented:
Applied all the review notes

AThousandShips commented:
You ignored quite a bit of those notes, they were relevant details like comment style and code formatting, it's important to pay attention to the details

LunaCapra commented:
That's my bad, I'll go over them again

AThousandShips commented:
Realized you had used mixed indentation in your code so updated my suggestion to fix that

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2391,8 +2419,51 @@ void fragment_shader(in SceneData scene_data) {
+	float rd = acos(cos_b);
+	float area = 0.0;
+
+	if (rd <= max(r1, r2) - min(r1, r2))
+	{
+		// One cone is enclosed in the other.
+		area = M_TAU - M_TAU * max(cos_a_v, cos_a_s);
+	}
+	else if (rd >= r1 + r2)
+	{
+		// No intersection
+		area = 0.0;
+	}
+	else
+	{
+		float diff = abs(r1 - r2);
+		float den = r1 + r2 - diff;
+		float x = 1.0 - clamp((rd - diff) / den, 0.0, 1.0);
+		area = smoothstep(0.0, 1.0, x);
+		area *= M_TAU - M_TAU * max(cos_a_v, cos_a_s);
+	}
```
```suggestion
	if (rd <= max(r1, r2) - min(r1, r2)) {
		// One cone is enclosed in the other.
		area = M_TAU - M_TAU * max(cos_a_v, cos_a_s);
	} else if (rd >= r1 + r2) {
		// No intersection.
		area = 0.0;
	} else {
		float diff = abs(r1 - r2);
		float den = r1 + r2 - diff;
		float x = 1.0 - clamp((rd - diff) / den, 0.0, 1.0);
		area = smoothstep(0.0, 1.0, x);
		area *= M_TAU - M_TAU * max(cos_a_v, cos_a_s);
	}
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2391,8 +2419,51 @@ void fragment_shader(in SceneData scene_data) {
+	// Apply cone to cone intersection with cosine weighted assumption:
+	// https://blog.selfshadow.com/publications/s2016-shading-course/activision/s2016_pbs_activision_occlusion.pdf
+	float cos_a_v = sqrt(1.0 - ao);
+	float limited_roughness = max(roughness, 0.01); // Avoid artifacts at really low roughness
```
```suggestion
	float limited_roughness = max(roughness, 0.01); // Avoid artifacts at really low roughness.
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2391,8 +2419,51 @@ void fragment_shader(in SceneData scene_data) {
+
+	float specular_occlusion = area / (M_TAU * (1.0 - cos_a_s));
+	indirect_specular_light *= ao_base * specular_occlusion;
+else // BENT_NORMAL_MAP_USED
```
```suggestion
#else // BENT_NORMAL_MAP_USED
```

LunaCapra commented:
Apologies about all the issues with the shader file formatting
This is my first time working on these

AThousandShips commented:
It's understandable, that's what we're here for :)

jcostello commented:
Why the effect is so different in mobile than forward?

LunaCapra commented:
> Why the effect is so different in mobile than forward?

It's a simpler calculation than forward+
I wouldn't mind putting the same calculation as forward+ into the mobile shader but it might perform badly with the cos and acos calls

Calinou commented on doc/classes/BaseMaterial3D.xml:
```diff
@@ -120,6 +120,15 @@
+			Texture used to specify the bent normal at a given pixel. The [member bent_normal_texture] only uses the red and green channels; the blue and alpha channels are ignored. The normal read from [member bent_normal_texture] is oriented around the surface normal provided by the [Mesh].
+			[b]Note:[/b] The mesh must have both normals and tangents defined in its vertex data. Otherwise, the normal map won't render correctly and will only appear to darken the whole surface. If creating geometry with [SurfaceTool], you can use [method SurfaceTool.generate_normals] and [method SurfaceTool.generate_tangents] to automatically generate normals and tangents respectively.
+			[b]Note:[/b] Godot expects the normal map to use X+, Y+, and Z+ coordinates. See [url=http://wiki.polycount.com/wiki/Normal_Map_Technical_Details#Common_Swizzle_Coordinates]this page[/url] for a comparison of normal map coordinates expected by popular engines.
+			[b]Note:[/b] The normals must be cosine weighted in order to function correctly.
```
"in order to" is redundant:

```suggestion
			[b]Note:[/b] The normals must be cosine weighted to function correctly.
```

Additionally, it may be useful to point to a resource that tells you how to generate bent normal maps. I assume these can't be generated from existing normal map textures though.

LunaCapra replied:
Yeah they are separately authored.
I will see if I can find a good resource.

LunaCapra replied:
I'm pushing the suggested changes for now as I can't find particularly good resources on bent normal generation.
For reference though:
Programs I know can bake them:

- Substance Painter
- Substance Designer
- xNormal (Unsure how exactly this program works)
- Blender (with the [right setup](https://blender.stackexchange.com/questions/157079/how-do-i-bake-a-bent-normal-map-in-blender))

Calinou reviewed (approved):
Tested locally, it works as expected. This change does a lot to make objects feel more grounded :slightly_smiling_face: 

All screenshots are Forward+ with SDFGI, SSAO, SSIL and glow enabled. No bent normal textures were used; I'm only testing the specular occlusion approximation.

There seems to be a small performance impact from this change, going from 360 FPS to 358 FPS in 4K on a RTX 4090. We should test this on integrated graphics as well to make sure it doesn't decrease performance too much there.

I've tested this change in Mobile as well but can't spot any visual difference there.

**Testing project:** [test_specular_occlusion.zip](https://github.com/godotengine/godot/files/14807331/test_specular_occlusion.zip)

### Metallic material

`master` | This PR
-|-
![Screenshot_20240329_182759_master png webp](https://github.com/godotengine/godot/assets/180032/adbd1d01-1994-440f-bd0b-4be077fa1ebd) | ![Screenshot_20240329_182747_with_specular_ao png webp](https://github.com/godotengine/godot/assets/180032/1c7134a0-ab09-4d52-8dc9-763eb49d93bc)

### Dielectric material

*Press <kbd>Space</kbd> in the testing project to change the material on the goblets.*

`master` | This PR
-|-
![Screenshot_20240329_183228_master png webp](https://github.com/godotengine/godot/assets/180032/afb14188-19e1-4364-9d52-3e7b9ad205cd) | ![Screenshot_20240329_183238_with_specular_ao png webp](https://github.com/godotengine/godot/assets/180032/ce6bda02-13d4-4409-a599-fc94cc3115e9)

### Metallic material 2

`master` | This PR
-|-
![Screenshot_20240329_184516_master png webp](https://github.com/godotengine/godot/assets/180032/0f73f923-a16a-42a7-8a8f-6526b4399e5b) | ![Screenshot_20240329_184502_with_specular_ao png webp](https://github.com/godotengine/godot/assets/180032/5e4f9b4a-eeb9-46a9-ab8c-79e2a498febc)


### Metallic material 2 with OmniLight3D

`master` | This PR
-|-
![Screenshot_20240329_184731_master png webp](https://github.com/godotengine/godot/assets/180032/f88cb649-6534-40a2-8d8c-924881ea1049) | ![Screenshot_20240329_184712_with_specular_ao png webp](https://github.com/godotengine/godot/assets/180032/eee681d2-b828-465c-a1e5-bbfe5d2104c4)

LunaCapra commented:
I fixed an error in using the direct light affect slider modified AO in the bent normal specular occlusion calculation instead of the baseline AO. 
This is now accurate to the paper referenced in the code.

| Before | After |
|--------|--------|
| ![image](https://github.com/godotengine/godot/assets/50772474/9aa9d2ae-fb11-4ac8-8042-d12f9da84fdc) | ![image](https://github.com/godotengine/godot/assets/50772474/ae40b939-31ed-4de9-98e0-8fb61925dcaa) |

LunaCapra commented:
I added a comment linking a PDF of the paper mentioned for spherical cap intersections.
This paper was only briefly mentioned in the main paper so I thought it'd be a good addition.

LunaCapra commented:
I created a [small scene](https://github.com/LunaCapra/godot-bent-normals-test/releases/download/1.0.0/BentNormalTest.zip) with CC0 models from polyhaven that I also generated bent normals for.
This should make testing this PR easier.
![image](https://github.com/godotengine/godot/assets/50772474/bc5f6db4-3d75-4f8a-999b-b789d90992d0)

LunaCapra committed:
Add support for bent normals for indirect lighting and specular occlusion

Mickeon commented on doc/classes/BaseMaterial3D.xml:
```diff
@@ -120,6 +120,15 @@
+			Texture used to specify the bent normal at a given pixel. The [member bent_normal_texture] only uses the red and green channels; the blue and alpha channels are ignored. The normal read from [member bent_normal_texture] is oriented around the surface normal provided by the [Mesh].
+			[b]Note:[/b] The mesh must have both normals and tangents defined in its vertex data. Otherwise, the normal map won't render correctly and will only appear to darken the whole surface. If creating geometry with [SurfaceTool], you can use [method SurfaceTool.generate_normals] and [method SurfaceTool.generate_tangents] to automatically generate normals and tangents respectively.
+			[b]Note:[/b] Godot expects the normal map to use X+, Y+, and Z+ coordinates. See [url=http://wiki.polycount.com/wiki/Normal_Map_Technical_Details#Common_Swizzle_Coordinates]this page[/url] for a comparison of normal map coordinates expected by popular engines.
+			[b]Note:[/b] The normals must be cosine weighted to function correctly.
```
_On the surface level_ this note is rather meaningless. Either it should be elaborated upon or removed.

LunaCapra replied:
I'll add a more elaborate explanation when I next rebase the PR

LunaCapra commented:
I rebased the branch, updated the documentation to hopefully be clearer/more elaborate, and also changed the main post to have more general and easier to compare example images.

Arazalera commented:
Could it be better to also alternatively just multiply reflection color by ambient occlusion?

lander-vr commented:
Just tested this PR in some environments using lightmapGI, relying on SSAO for specular occlusion:

| Master | PR |
| - | - |
| ![image](https://github.com/user-attachments/assets/a27358a4-daa7-43f4-8c22-678653b7e14a) | ![image](https://github.com/user-attachments/assets/5b66241c-19c0-4b93-ae36-6e09d60cd324) |
| ![image](https://github.com/user-attachments/assets/b66188fc-d6b9-4196-b167-d6c44e28d130) | ![image](https://github.com/user-attachments/assets/3fef92b7-2436-423a-854b-1651e5abedd9) |

(Tile-room is a quick remake of the scene in this [Unity forum post](https://discussions.unity.com/t/specular-occlusion/807318))

I imagine this PR would benefit _hugely_ from a reworked SSAO down the line, as the current SSAO technique struggles with capturing depth and detail.

This would give stellar results used in conjunction with https://github.com/godotengine/godot/pull/86102.

LunaCapra commented:
With #106145 merged I adjusted the PR to support the new method of specular occlusion.
This means the specular occlusion part of bent normals can be disabled via the project settings or per material.

I also added a long missing visual shader output for bent normal maps.

Calinou commented:
> I imagine this PR would benefit hugely from a reworked SSAO down the line, as the current SSAO technique struggles with capturing depth and detail.

Can you give https://github.com/godotengine/godot/pull/101961 a try?

clayjohn commented on drivers/gles3/storage/material_storage.cpp:
```diff
@@ -1156,6 +1156,7 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["NORMAL"] = "normal";
 		actions.renames["NORMAL_MAP"] = "normal_map";
 		actions.renames["NORMAL_MAP_DEPTH"] = "normal_map_depth";
+		actions.renames["BENT_NORMAL_MAP"] = "bent_normal_map";
```
Not needed since this section is for CanvasItem shaders

clayjohn commented on drivers/gles3/storage/material_storage.cpp:
```diff
@@ -1190,6 +1191,7 @@ MaterialStorage::MaterialStorage() {
 		actions.usage_defines["SCREEN_PIXEL_SIZE"] = "@SCREEN_UV";
 		actions.usage_defines["NORMAL"] = "#define NORMAL_USED\n";
 		actions.usage_defines["NORMAL_MAP"] = "#define NORMAL_MAP_USED\n";
+		actions.usage_defines["BENT_NORMAL_MAP"] = "#define BENT_NORMAL_MAP_USED\n";
```
Same as above, this doesn't need to be defined for CanvasItem shaders

clayjohn commented on servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile.glsl:
```diff
@@ -1489,7 +1515,7 @@ void main() {
 #ifdef USE_VERTEX_LIGHTING
 	diffuse_light += diffuse_light_interp.rgb;
-	specular_light += specular_light_interp.rgb * f0;
+	indirect_specular_light += specular_light_interp.rgb * f0;
```
This and the next couple should be `direct_specular_light` since vertex lighting is all direct lighting

clayjohn commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl:
```diff
@@ -22,13 +23,13 @@
 #include "../scene_data_inc.glsl"
 
-#if !defined(MODE_RENDER_DEPTH) || defined(MODE_RENDER_MATERIAL) || defined(MODE_RENDER_SDF) || defined(MODE_RENDER_NORMAL_ROUGHNESS) || defined(MODE_RENDER_VOXEL_GI) || defined(TANGENT_USED) || defined(NORMAL_MAP_USED) || defined(LIGHT_ANISOTROPY_USED)
+#if !defined(MODE_RENDER_DEPTH) || defined(MODE_RENDER_MATERIAL) || defined(MODE_RENDER_SDF) || defined(MODE_RENDER_NORMAL_ROUGHNESS) || defined(MODE_RENDER_VOXEL_GI) || defined(TANGENT_USED) || defined(NORMAL_MAP_USED) || defined(BENT_NORMAL_MAP_USED) || defined(LIGHT_ANISOTROPY_USED)
```
We need this same change in scene_forward_clustered_mobile_inc.glsl

clayjohn commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
```diff
@@ -2030,7 +2084,7 @@ void fragment_shader(in SceneData scene_data) {
 #ifdef USE_VERTEX_LIGHTING
 	diffuse_light += diffuse_light_interp.rgb;
-	specular_light += specular_light_interp.rgb * f0;
+	indirect_specular_light += specular_light_interp.rgb * f0;
```
Should be direct_specular_light since vertex lighting is all direct

clayjohn commented on doc/classes/BaseMaterial3D.xml:
```diff
@@ -120,6 +120,15 @@
+			Texture used to specify the vector pointing in the average direction of incoming ambient light at a given pixel. The [member bent_normal_texture] only uses the red and green channels; the blue and alpha channels are ignored. The normal read from [member bent_normal_texture] is oriented around the surface normal provided by the [Mesh].
+			[b]Note:[/b] The mesh must have both normals and tangents defined in its vertex data. Otherwise, the bent normal effect won't render correctly. If creating geometry with [SurfaceTool], you can use [method SurfaceTool.generate_normals] and [method SurfaceTool.generate_tangents] to automatically generate normals and tangents respectively.
+			[b]Note:[/b] Godot expects the normal map to use X+, Y+, and Z+ coordinates. See [url=http://wiki.polycount.com/wiki/Normal_Map_Technical_Details#Common_Swizzle_Coordinates]this page[/url] for a comparison of normal map coordinates expected by popular engines.
+			[b]Note:[/b] The bent normal vectors must have a cosine weighted angle distribution. Otherwise, the resulting specular occlusion and indirect lighting will not look correct.
```
How can users ensure that their bent normal vectors maps use a cosine weighted angle distribution? Is this a setting in content creation applications like Blender?

clayjohn replied:
After posting this I noticed that I am the third person to comment on this particular line. So I'll add a bit of clarity why it is tripping me up. 

Documentation that highlights a potential problem should also give the user the information to solve that problem. For 99% of users "cosine weighted angle distribution" is meaningless jargon. They can't do anything with that information. Instead the documentation should focus on things that the user can control. 

To do that, we need to step back and think about what we are trying to communicate to the user here. I'm not familiar with baking bent normals. But my gut says that the message this should be conveying is "you need to use a proper program to calculate the bent normals, you can't just pop in a different normal map or a scaled normal map or something and expect this to work."

Unreal engine has a nice, but long, explanation https://dev.epicgames.com/documentation/en-us/unreal-engine/bent-normal-maps-in-unreal-engine Something like this would be good to add in https://docs.godotengine.org/en/latest/tutorials/3d/standard_material_3d.html

LunaCapra replied:
> How can users ensure that their bent normal vectors maps use a cosine weighted angle distribution? Is this a setting in content creation applications like Blender?

Yes, it is. It's similar to ambient occlusion as it can also have a linear or cosine weighted distribution when baking.
I adjusted the note in `BaseMaterial3D`'s docs to specify that this is needed when baking the texture.

And on the note of the  `StandardMaterial3D` documentation page: It seems it also still needs a description of the "Disable Specular Occlusion" property under shading.

clayjohn replied:
> And on the note of the StandardMaterial3D documentation page: It seems it also still needs a description of the "Disable Specular Occlusion" property under shading.

Yes, that would be good to add as well

clayjohn reviewed (changes requested):
Amazing work! This is looking really good and the implementation is super clean! I'm glad it fits in so nicely with the recent specular occlusion merge. I left a few comments for some issues I spotted, but there is nothing seriously problematic. We should be able to merge soon

clayjohn reviewed (approved):
Looks fantastic! I tested it locally using the provided project and it worked great!

I ran it through all three backends and checked materials with bent normal maps provided and those without and noticed no issues. 

So I'd say this is ready for merging!

Repiteo commented:
Thanks!

Calinou commented:
Out of curiosity, is it planned to add bent normal support to Compatibility later on?

mechalynx commented:
I know this is already merged, I only just read about it from the blog post but since testing on internal graphics was requested I decided to test it anyway, since I didn't see anyone reporting here about that, just to be sure.

System info:
> Godot v4.5.dev (2334519a4) - Windows 10 (build 19043) - Multi-window, 2 monitors - Vulkan (Forward+) - integrated AMD Radeon(TM) Graphics (Advanced Micro Devices, Inc.; 31.0.21921.1000) - AMD Ryzen 5 5600G with Radeon Graphics (12 threads)

I've got 4GB allocated to the internal gpu and about half of it was in use during tests so that shouldn't affect frame times between the two.

Testing the given project with this PR (from the artifacts, I didn't compile locally), I get about `5.10ms/195fps` in the editor, though it did give me some errors because it couldn't find .jpg versions of the textures but it seemed fine using the .png versions, so I just left it as is. GPU usage is about 52% with the editor focused.

Trying it with v4.5.dev.custom_build [2cde9292c] which I had previously compiled locally, there is no noticable difference in performance as far as I can tell, `5.10-5.25ms/190-195fps` with the editor focused.  GPU usage was more like 53% but even though it was stable there, I'm not sure if that would necessarily mean anything - it's also an older commit.

However, looking at comparison screenshots between the two, I can't tell any difference. Not sure if I'm doing something wrong, SSAO is enabled in the world environment in the scene and is working (toggling it and changing settings makes a visible difference).

screenshot from locally built [2cde9292c]:
![image](https://github.com/user-attachments/assets/9cc5121f-8fed-4e5b-bf41-0ca922ffb7be)
![image](https://github.com/user-attachments/assets/5dc8fe3b-ed5a-403e-aef5-0532fd1dd454)

screenshot from this PR:
![image](https://github.com/user-attachments/assets/6dce6aee-d38c-48cc-81fe-fc2cb32b1258)
![image](https://github.com/user-attachments/assets/8294f943-180e-4b9d-852c-644914b3f8e3)
